lib/native/mqtt.ts

```diff
@@ -54,30 +48,67 @@ export interface ConnectionConfig {
     tls_ctx?: io.ClientTlsContext;
 }
 
-export class Connection extends NativeResource implements ResourceSafety.ResourceSafe {
-    public client: Client;
-    private encoder: TextEncoder;
-    private config: ConnectionConfig
-
-    constructor(client: Client, config: ConnectionConfig, on_connection_interrupted?: (error_code: number) => void, on_connection_resumed?: (return_code: number, session_present: boolean) => void) {
-        super(crt_native.mqtt_client_connection_new(client.native_handle(), on_connection_interrupted, on_connection_resumed));
-        
-        this.client = client;
-        this.config = config;
-        this.encoder = new TextEncoder();
+export class MqttClientConnection extends NativeResourceMixin(BufferedEventEmitter) {
+    private encoder = new TextEncoder();
+    readonly tls_ctx?: io.ClientTlsContext; // this keeps the tls_ctx alive beyond the life of the connection
+
+    constructor(readonly client: MqttClient, private config: MqttConnectionConfig) {
+        super();
+        this._super(crt_native.mqtt_client_connection_new(
+            client.native_handle(),
+            (error_code: number) => { this._on_connection_interrupted(error_code); },
+            (return_code: number, session_present: boolean) => { this._on_connection_resumed(return_code, session_present); })
+        );
+        this.tls_ctx = config.tls_ctx;
     }
 
-    close() {
+    private close() {
         crt_native.mqtt_client_connection_close(this.native_handle());
     }
 
+    /** Emitted when the connection is ready and is about to start sending response data */
+    on(event: 'connect', listener: (session_present: boolean) => void): this;
+
+    /** Emitted when connection has disconnected sucessfully. */
+    on(event: 'disconnect', listener: () => void): this;
+
+    /**
+     * Emitted when an error occurs
+     * @param error - A CrtError containing the error that occurred
+     */
+    on(event: 'error', listener: (error: CrtError) => void): this;
+
+    /**
+     * Emitted when the connection is dropped unexpectedly. The error will contain the error
+     * code and message.
+     */
+    on(event: 'interrupt', listener: (error: CrtError) => void): this;
+
+    /**
+     * Emitted when the connection reconnects. Only triggers on connections after the initial one.
+     */
+    on(event: 'resume', listener: (return_code: number, session_present: boolean) => void): this;
```

Comment:
why does 'resume' pass `return_code` but 'connect' doesn't?

Reply:
If you connected, the only possible return code is 0 AFAIK. So we can either leave it out, or just pass 0s to both.

Reply:
you know what. python has the same thing.
I don't know what's up with return_code either
not sure exactly what state you're in if that's non-zero